dNR: stop serializing rule identifiers with their actions and introduce a new action type instead

elijahsawyers opened the pull request:
#### f4f5e767d0b6eec1351fa9fb1f1224287a668edc
<pre>
dNR: stop serializing rule identifiers with their actions and introduce a new action type instead
<a href="https://bugs.webkit.org/show_bug.cgi?id=297338">https://bugs.webkit.org/show_bug.cgi?id=297338</a>
<a href="https://rdar.apple.com/157879177">rdar://157879177</a>

Reviewed by Timothy Hatcher and Alex Christensen.

This patch stops serializing rule identifiers with their actions. Previously, the goal was to have
a rule&apos;s identifier be serialized with its action, and we&apos;d write a bit to the metadata to indicate
if serialized actions included identifiers. This meant that every rule had to have an identifier.

Now, during parsing, if a rule has an identifier, we create another rule that has the same trigger
but the new action type, ReportIdentifierAction. If we encounter this action while processing a load,
we publish the notification. Splitting out a rule&apos;s identifier into a separate rule has the benefit
of not requiring all rules to have an identifier, as well as being able to combine actions during
serialization. However, the downside is that we&apos;re consuming more memory than the aforementioned
approach.

Also, per the recommendation of @achristensen07, I removed the compile time guards from the content
extensions code. However, I did leave them for the web extensions code, as it&apos;s common practice.

Modified the pre-existing tests to verify this change.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/contentextensions/ContentExtension.cpp:
(WebCore::ContentExtensions::ContentExtension::findFirstIgnoreRule const):
* Source/WebCore/contentextensions/ContentExtensionActions.h:
* Source/WebCore/contentextensions/ContentExtensionCompiler.cpp:
(WebCore::ContentExtensions::serializeActions):
* Source/WebCore/contentextensions/ContentExtensionError.cpp:
(WebCore::ContentExtensions::contentExtensionErrorCategory):
* Source/WebCore/contentextensions/ContentExtensionError.h:
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::loadRule):
(WebCore::ContentExtensions::loadRuleIdentifier):
(WebCore::ContentExtensions::loadEncodedRules):
* Source/WebCore/contentextensions/ContentExtensionRule.cpp:
(WebCore::ContentExtensions::ContentExtensionRule::ContentExtensionRule):
(WebCore::ContentExtensions::DeserializedAction::deserialize):
* Source/WebCore/contentextensions/ContentExtensionRule.h:
(WebCore::ContentExtensions::ContentExtensionRule::isolatedCopy):
(WebCore::ContentExtensions::ContentExtensionRule::identifier const): Deleted.
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForPingLoad):
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForResourceMonitoring):
* Source/WebCore/contentextensions/ContentRuleListMatchedRule.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::contentRuleListNotification):
(WebCore::ChromeClient::contentRuleListMatchedRule):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::contentRuleListNotification):
(API::NavigationClient::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::contentRuleListNotification):
(WebKit::NavigationState::NavigationClient::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::contentRuleListNotification):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::contentRuleListMatchedRule):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::TEST_F(ContentExtensionTest, InvalidJSON)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, ReportIdentifierAction)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, RuleIdentifiers)): Deleted.

Canonical link: <a href="https://commits.webkit.org/299028@main">https://commits.webkit.org/299028@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2b825f6b6f7810ea1a61c80b599984dacf067b0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117498 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37174 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27797 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123603 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69504 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7ff3db45-4b22-4408-9289-ddbae85a73e1) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119376 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37868 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45758 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89177 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/44985 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e5b4e1fe-76e9-4dcc-a80a-09fac1822e0c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120450 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30157 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105364 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69685 "Found 40 new API test failures: /WPE/TestDownloads:beforeAll, /WPE/TestWebKitPolicyClient:beforeAll, /WPE/TestWebKitSecurityOrigin:beforeAll, /WPE/TestEditor:beforeAll, /WPE/TestUIClient:beforeAll, /TestWebCore:GStreamerTest.harnessDecodeMP4Video, /WPE/TestFrame:beforeAll, /WPE/TestBackForwardList:beforeAll, /WPEPlatform/TestDisplayWayland:/wpe-platform/DisplayWayland/available-input-devices, /WPEPlatform/TestDisplayDefault:/wpe-platform/Display/default ... (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9d0aad1f-1628-44f9-abc7-3641636818f9) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29217 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23481 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67275 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/109608 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99571 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23664 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126723 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/116007 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44401 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33404 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97848 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44759 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101595 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97636 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43009 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20951 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40758 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18760 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44272 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49947 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144705 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43729 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37244 "Found 1 new JSC binary failure: testapi, Found 20033 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default, ChakraCore.yaml/ChakraCore/test/Function/StackArgsWithFormals.js.default, ChakraCore.yaml/ChakraCore/test/Function/apply.js.default, ChakraCore.yaml/ChakraCore/test/Function/apply3.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47078 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45422 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/Shared/WebCompiledContentRuleListData.h:
```diff
@@ -63,6 +74,9 @@ class WebCompiledContentRuleListData {
     uint64_t frameURLFiltersBytecodeOffset { 0 };
     uint64_t frameURLFiltersBytecodeSize { 0 };
+#if ENABLE(DNR_ON_RULE_MATCHED_DEBUG)
+    uint32_t actionsIncludeIdentifiers { false };
```
Why is this not `bool`?

xeenon commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -115,7 +115,8 @@ struct ContentRuleListMetaData {
     uint64_t topURLFiltersBytecodeSize { 0 };
     uint64_t frameURLFiltersBytecodeSize { 0 };
-    uint32_t unused32bits { false };
+    uint16_t actionsIncludeIdentifiers { false };
```
Ditto.

xeenon commented on Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
```diff
@@ -392,6 +410,9 @@ static Expected<MappedData, std::error_code> compiledToFile(WTF::String&& json,
         size_t m_topURLFiltersBytecodeWritten { 0 };
         size_t m_frameURLFiltersBytecodeWritten { 0 };
+#if ENABLE(DNR_ON_RULE_MATCHED_DEBUG)
+        bool m_actionsIncludeIdentifiers { false };
```
It is here.

xeenon commented on Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
```diff
@@ -69,6 +69,10 @@ struct CompiledContentExtensionData {
     Vector<ContentExtensions::DFABytecode> frameURLFilters;
+
+#if ENABLE(DNR_ON_RULE_MATCHED_DEBUG)
+    bool actionsIncludeIdentifiers;
```
And here.

achristensen07 reviewed (changes requested):
This all looks messy, and it's unclear what kind of identifiers you're talking about, and it's unclear why there is a compile time flag for all of this, and it's unclear what is motivating this change.

achristensen07 commented on Source/WebCore/contentextensions/ContentExtensionParser.cpp:
```diff
@@ -297,19 +297,24 @@ static std::optional<Expected<ContentExtensionRule, std::error_code>> loadRule(c
+    if (!trigger.has_value())
+        return makeUnexpected(trigger.error());
+
+    return { { { WTFMove(trigger.value()), Action { ReportIdentifierAction { static_cast<uint32_t>(identifier.value()) } } } } };
```
If we went with actually serializing the identifier from the JSON we would either have to serialize more than 32 bits, limit the max identifier to UINT_MAX, or truncate and lose some values.  Why don't we just serialize the index of the rule in the original JSON array instead and go get the identifier and other action properties later using that index?

elijahsawyers replied:
Spoke offline. Going to update the serialized identifier to be a `double`.

achristensen07 commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -319,16 +316,14 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForLoad(
+            }, [&] (const ReportIdentifierAction& reportIdentifierAction) {
+                // FIXME: <rdar://157880177> Include the rest of the parameters on the ContentRuleListMatchedRule struct.
+                ContentRuleListMatchedRule matchedRule;
+                matchedRule.request.url = url.string();
```
It's kind of an anti-pattern to default construct an object then populate its members one by one.  Can we just pass everything to the constructor instead?  That way we wouldn't need to have all the members std::optional.  Also, this would probably be good for a separate PR since it's not really related to this.

elijahsawyers replied:
Yeah, I'm going to tackle that with the FIXME radar above.

elijahsawyers committed:
dNR: stop serializing rule identifiers with their actions and introduce a new action type instead
https://bugs.webkit.org/show_bug.cgi?id=297338
rdar://157879177

Reviewed by Timothy Hatcher and Alex Christensen.

This patch stops serializing rule identifiers with their actions. Previously, the goal was to have
a rule's identifier be serialized with its action, and we'd write a bit to the metadata to indicate
if serialized actions included identifiers. This meant that every rule had to have an identifier.

Now, during parsing, if a rule has an identifier, we create another rule that has the same trigger
but the new action type, ReportIdentifierAction. If we encounter this action while processing a load,
we publish the notification. Splitting out a rule's identifier into a separate rule has the benefit
of not requiring all rules to have an identifier, as well as being able to combine actions during
serialization. However, the downside is that we're consuming more memory than the aforementioned
approach.

Also, per the recommendation of @achristensen07, I removed the compile time guards from the content
extensions code. However, I did leave them for the web extensions code, as it's common practice.

Modified the pre-existing tests to verify this change.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/contentextensions/ContentExtension.cpp:
(WebCore::ContentExtensions::ContentExtension::findFirstIgnoreRule const):
* Source/WebCore/contentextensions/ContentExtensionActions.h:
* Source/WebCore/contentextensions/ContentExtensionCompiler.cpp:
(WebCore::ContentExtensions::serializeActions):
* Source/WebCore/contentextensions/ContentExtensionError.cpp:
(WebCore::ContentExtensions::contentExtensionErrorCategory):
* Source/WebCore/contentextensions/ContentExtensionError.h:
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
(WebCore::ContentExtensions::loadRule):
(WebCore::ContentExtensions::loadRuleIdentifier):
(WebCore::ContentExtensions::loadEncodedRules):
* Source/WebCore/contentextensions/ContentExtensionRule.cpp:
(WebCore::ContentExtensions::ContentExtensionRule::ContentExtensionRule):
(WebCore::ContentExtensions::DeserializedAction::deserialize):
* Source/WebCore/contentextensions/ContentExtensionRule.h:
(WebCore::ContentExtensions::ContentExtensionRule::isolatedCopy):
(WebCore::ContentExtensions::ContentExtensionRule::identifier const): Deleted.
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForPingLoad):
(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForResourceMonitoring):
* Source/WebCore/contentextensions/ContentRuleListMatchedRule.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::contentRuleListNotification):
(WebCore::ChromeClient::contentRuleListMatchedRule):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/APINavigationClient.h:
(API::NavigationClient::contentRuleListNotification):
(API::NavigationClient::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/Cocoa/NavigationState.h:
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::NavigationState::NavigationClient::contentRuleListNotification):
(WebKit::NavigationState::NavigationClient::contentRuleListMatchedRule):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::contentRuleListNotification):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::contentRuleListMatchedRule):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Tools/TestWebKitAPI/Tests/WebCore/ContentExtensions.cpp:
(TestWebKitAPI::TEST_F(ContentExtensionTest, InvalidJSON)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, ReportIdentifierAction)):
(TestWebKitAPI::TEST_F(ContentExtensionTest, RuleIdentifiers)): Deleted.

Canonical link: https://commits.webkit.org/299028@main
